### DISCUSSION TITLE
Non-random names?

### DISCUSSION THREAD

**Discussion by seqizz** [orhun/rustypaste#216] [Q&A]
I had another small use case: My reaction folder can be enhanced by rustypaste 😼 But needs one small thing I couldn't find on the readme.. Can I decide resulting filename while uploading the file? No random or anything.

**Comment by orhun** [comment]
Yes, you can do this.

First, disable random file names:

```toml
[paste]
# random_url = { type = "petname", words = 2, separator = "-" }
# random_url = { type = "alphanumeric", length = 8 }
```

Now you can give a custom file name to the uploaded file with using the form field:

```bash
$ curl -F "file=@x.txt;filename=wow.nice" http://localhost:8000
http://localhost:8000/wow.nice
```

[`rpaste`](https://github.com/orhun/rustypaste-cli) does not have an argument to support this though. Maybe we can add `rpaste --file-name` or something. Do you use `rpaste`, what do you think?

**Reply by seqizz** [reply]
Hmm, random by default is a useful thing though. Might be cool to have an rpaste argument like your suggestion to bypass this, resulting the server to ignore the `random_url` setting temporarily, with an auth token of course. :thinking:

**Reply by orhun** [reply]
Well, we don't need to interact with the server at all. `--file-name` argument just should set the `filename` argument for the multipart request regardless of the server has random URLs set or not.

**Reply by seqizz** [reply]
Then not sure if I understand, I tested on my server which has:
```
random_url = { enabled = true, type = "petname", words = 2, separator = "-" }
```

by creating following test script:
```
function testpaste() {
  curl -F "file=@$1;filename=testing.jpg" -H "Authorization: mytoken" "https://servername"
}
```

And server did not reply with `testing.jpg`, it stored with another random string :thinking:

**Reply by orhun** [reply]
Ah, I see. You still want to enable random file names but also have an option to set a file name as well. (My brain was slow when I wrote the previous reply, sorry.)

I think we can support this by checking:

- When only `file` argument is given via multipart request, `filename` is the same as `file`. In this case, disable/enable random file names as usual.
- If both `file` and `filename` arguments are given and if they are not the same, use the given `filename` to store the file.

Alternatively, we can add a `--file-name` argument to `rpaste` so you can actually send a custom `filename` in the multipart request.

**Reply by seqizz** [reply]
```
If both file and filename arguments are given and if they are not the same, use the given filename to store the file.
```
There is a tiny nitpick question though: If you'd like server-side to be able to enforce random names for any reason (like security 😬) there has to be another option, disallowing this by default. E.g. `allow_custom_filename = false` either on root or inside `random_url` dict value.

**Reply by orhun** [reply]
Good point, I think we can simply do:

```toml
random_url = { type = "alphanumeric", length = 8, enforce = true }
```

**Reply by orhun** [reply]
I think the requirements for the implementation is pretty clear now, can you create a tracking issue so that we can get this rolling? 🐻
